File: lib/java/src/main/java/com/dianping/cat/message/internal/DefaultMetric.java
package com.dianping.cat.message.internal;

import com.dianping.cat.message.Metric;
import com.dianping.cat.message.spi.MessageManager;

public class DefaultMetric extends AbstractMessage implements Metric {
    private MessageManager manager;

    public DefaultMetric(String type, String name) {
        super(type, name);
    }

    public DefaultMetric(String type, String name, MessageManager manager) {
        super(type, name);

        this.manager = manager;
    }

    @Override
    public void complete() {
        setCompleted(true);

        if (manager != null) {
            manager.add(this);
        }
    }
}

File: lib/csharp/src/CatClient/Message/Internals/DefaultMetric.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Org.Unidal.Cat.Message.Spi;

namespace Org.Unidal.Cat.Message.Internals
{
    public class DefaultMetric : AbstractMessage, IMetric
    {
        private IMessageManager _mManager;

        public DefaultMetric(String type, String name) : base(type, name)
        {
        }

        public DefaultMetric(string type, string name, IMessageManager manager)
            : base(type, name)
        {
            _mManager = manager;
        }

        public override void Complete()
        {
            try
            {
                base.Complete();

                if (_mManager != null && _mManager.ThreadLocalMessageTree.Message == null)
                {
                    _mManager.Add(this);
                }
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
            }
        }
    }
}
